### ISSUE TITLE
Profiler enhancements

### ISSUE THREAD

**Issue by mmcgr** [souffle-lang/souffle#634]
We've been working on enhancing the profiler. firstly to add more information by tracking memory use, more accurate timing, and adding frequency statistics to each rule. With these you can get a better idea of which rules are more productive, how much memory is being used at any time, and what is using the memory. With some slight polishing these will be ready to add to the logs, and soon the CLI profiler will visualise the new data.

We would also like to enhance the profiler interface to make it more useful. We were thinking of making it possible to attach to a running souffle instance, possibly on a different machine, This would be similar to the way the profiler can be used now, via the logs, but more directly and with a lower impact on performance.

**Would these be useful enhancements? How is the profiler used now?** It wasn't until we changed the logging that we discovered how well used the profiling output is, so input on possible enhancements would be quite useful.

**Comment by yanniss** [comment]
Here are my thoughts on the profiler. 
(BTW, I did not notice the #628 changes as they happened--I just saw the comment on #628 and chimed in. But just a week earlier I had done a new installation of souffle--on a Facebook datacenter machine--and if I had gotten the #628 changes, I'd have noticed.)

The old profiler worked just fine for me. In fact, it was wonderful! I already got a good idea of rule frequency, memory consumption, rule cost, etc. I didn't see a need for more accurate tracking of memory use, timing, or frequency statistics. Only in very marginal cases I've needed to look at the generated C++ code, and even then I almost never got any more insights.

So, for me the desired enhancements of the old profiler would have been very pedestrian stuff:
 - avoid the double compilation times just for adding profiling! Not sure which C++ library caused this since I never looked into it, but it was extreme!
 - work with g++ < v.7 (I think that one needs at least v.7 for correct profiling? Compilation succeeds, but the profiler's output is near-empty for g++ 6.4).

Basically, I use profiling in two ways:
 a) to ensure that rules are efficient (it's too easy to write vastly inefficient rules). Live mode is essential for this!
 b) to ensure that rules are correct (check statistics and see if a newly-written rule produced zero tuples--the best way to debug! Or compare two profiles to see which rules produced different counts after a refactoring--we caught a very nasty bug this way recently).

Finally, I've very much liked that profiling was lightweight and portable. At some point in the past when the profiler was broken, I wrote a Java class that parsed the profile events and gave me a poor-man's profile.

So, overall, my feedback would be "keep it simple, lightweight, and highly robust, please". Others may want more bells and whistles, though.

**Comment by mmcgr** [comment]
#661 covers the enhancements now planned
